// gupshup-otp/script/index.js
// var botConfig = {
//     apikey: "ac343e77e3944eb1ce4ce0c74dfefbad",
//     botname: "simpleBot",
// }
// var otpConfig = {
//     accountid: "2000043378",
//     token: "webar00",
//     customMsg: botConfig.botname+" : Hello, your otp is %code%",
//     mujahid: "9702347733",
//     me: "9975303374",
//     meJio: "8779560349",
//     agnibha: "7710004077",
//     kamlesh: "9833646736"
//     // "botname":"somebot",
//     // "otpkey":"b4077ef37d95ee9984a89b4aef2b297e"
// }
// var config;
// config = Object.assign({}, otpConfig, botConfig);
// var otp = new (require('gupshup-otp'))(config,function(){
//     console.log("########## I am the init function ##############");
// });
// otp = new otp(config);

var executor = require("bot-script").executor;
function MessageHandler(context, event) {
    var options = {};
    options.current_dir = __dirname;
    options.start_section = "default.main";
    options.success = function (opm) {
        console.log(opm);
        context.sendResponse(JSON.stringify(opm));
    };
    options.error = function (err) {
        console.log(err.stack);
        context.sendResponse("Some Error Occurred");
    };
    executor.execute(options, event, context);
}
function MessageHandlerOTP(context, event) {
    var msg = event.message.toLowerCase();
    if (msg.split(" ")[0] == "verify") {
        context.console.log(msg,context.simpledb.roomleveldata.phone.number)
        otp.verifyOtp({
            phone:context.simpledb.roomleveldata.phone.number,
            code: msg.split(" ")[1]
        }, function (resp) {
            context.console.log("Response from "+resp);
            if(!resp.status.error)
                context.simpledb.roomleveldata.phone.verified=true;
            context.sendResponse(resp);
        });
    }
    else {
        // handleOtp(context,event);
        var number = '91' + config.meJio;
        otp.sendOtp(number, function (resp) {
            console.log('sent otp');
            console.log(resp)
            context.console.log(resp);
            context.simpledb.roomleveldata.phone = {number:number,verified:false};
            context.simpledb.saveData(function(){
                context.sendResponse(resp);
            });
        })
    }
}

function EventHandler(context, event) {
    context.simpledb.roomleveldata = {};
    MessageHandler(context, event);
}

function handleOtp(context, event) {
    var msg = event.message.toLowerCase();
    var state = msg.split(" ")[0];
    var number = msg.split(" ")[1];
    if (state == "otp") {
        var phone = "91" + number;
        var form = "?phone=" + number + "&key=b4077ef37d95ee9984a89b4aef2b297e&format=json";
        try {
            context.simplehttp.makeGet("http://enterprise.smsgupshup.com/apps/TwoFactorAuth/incoming.php" + form, {},
                function (cont, evt) {
                    var resp = JSON.parse(evt.getresp);
                    context.simpledb.roomleveldata.phone = number;
                    context.simpledb.saveData(function () {
                        context.sendResponse("Otp sent on your mobile " + number + ":\n" + resp.response);
                    })
                })
        } catch (err) {
            context.sendResponse("Error", err);
        }
    }
    else if (state == "verify") {
        try {
            var phone = "91" + context.simpledb.roomleveldata.phone;
            var form = "?phone=" + phone + "&key=b4077ef37d95ee9984a89b4aef2b297e&format=json&code=" + number;
            context.simplehttp.makeGet("http://enterprise.smsgupshup.com/apps/TwoFactorAuth/incoming.php" + form, {},
                function (cont, evt) {
                    var resp = JSON.parse(evt.getresp);
                    context.sendResponse("Response " + number + ": " + phone + "\n" + resp.response);
                })
        } catch (err) {
            context.sendReponse("Error", err);
        }
    }
    else {
        context.sendResponse(state);
    }
}

function HttpResponseHandler(context, event) {
    if (event.geturl === "http://ip-api.com/json")
        context.sendResponse('This is response from http \n' + JSON.stringify(event.getresp, null, '\t'));
}

function DbGetHandler(context, event) {
    context.sendResponse("testdbput keyword was last sent by:" + JSON.stringify(event.dbval));
}

function DbPutHandler(context, event) {
    context.sendResponse("testdbput keyword was last sent by:" + JSON.stringify(event.dbval));
}

function HttpEndpointHandler(context, event) {
    context.sendResponse('This is response from http \n' + JSON.stringify(event, null, '\t'));
}

function LocationHandler(context, event) {
    context.sendResponse("Got location");
};

exports.onMessage = MessageHandler;
exports.onEvent = EventHandler;
exports.onHttpResponse = HttpResponseHandler;
exports.onDbGet = DbGetHandler;
exports.onDbPut = DbPutHandler;
if (typeof LocationHandler == 'function') {
    exports.onLocation = LocationHandler;
}
if (typeof HttpEndpointHandler == 'function') {
    exports.onHttpEndpoint = HttpEndpointHandler;
}
